salt/_modules/metalk8s_service_configuration.py

```diff
@@ -0,0 +1,105 @@
+'''Metalk8s cluster and service configuration utility.'''
+
+import logging
+import yaml
+
+from salt.exceptions import CommandExecutionError
+
+
+log = logging.getLogger(__name__)
+
+__virtualname__ = 'metalk8s_service_configuration'
+
+
+def __virtual__():
+    return __virtualname__
+
+
+def get_service_conf(
+    namespace,
+    configmap_name,
+    kind=None,
+    apiVersion=None,
+    **kwargs
+):
+    """Reads a ConfigMap from the name specified
+
+    Arguments:
+        configmap_name: the ConfigMap name
+        namespace: the Namespace where the ConfigMap is stored
+
+    Returns:
+        A dict of a specific service configuration
+
+    CLI Examples:
+
+    .. code-block:: bash
+
+        salt-call metalk8s_service_configuration.get_service_conf "metalk8s-monitoring" "metalk8s-prometheus-config"
+    """
+
+    if not configmap_name:
+        raise Exception(
+            'Expected a ConfigMap name but got {}'.format(configmap_name)
+        )
+
+    try:
+        manifest = __salt__[
+            'metalk8s_kubernetes.get_object'
+        ](
+            kind='ConfigMap',
+            apiVersion='v1',
+            namespace=namespace,
+            name=configmap_name
+        )
+    except ValueError as exc:
+        raise CommandExecutionError(
+            'Failed to read ConfigMap object {}: {!s}'
+            .format(configmap_name, exc)
+        )
+
+    if manifest is None:
+        raise CommandExecutionError(
+            'Expected ConfigMap object but got {}'.format(manifest)
+        )
+
+    try:
+        conf_section = manifest.get('data', {}).get('config.yaml', {})
+        config = yaml.safe_load(conf_section) or {}
+    except yaml.YAMLError as exc:
+        raise CommandExecutionError(
+            'Invalid YAML format in Configmap {}: {!s}'.format(
+                configmap_name, exc
+            )
+        )
+    except Exception as exc:
+        raise CommandExecutionError(
+            'Failed loading `config.yaml` from Configmap {}: {!s}'.format(
+                configmap_name, exc
+            )
+        )
+    if not config:
+        raise CommandExecutionError(
+            'Expected `config.yaml` as yaml in the ConfigMap {} but got {}'
+            .format(configmap_name, config)
+        )
+
+    if apiVersion is None and kind is None:
+        return config
+
+    if config['apiVersion'] != apiVersion:
+        raise CommandExecutionError(
+            'Expected value {} for key apiVersion, got {}'.format(
+                config['apiVersion'], apiVersion
+            )
+        )
+
+    if config['kind'] != kind:
+        raise CommandExecutionError(
+            'Expected value {} for key kind, got {}'.format(
+                config['kind'], kind
+            )
+        )
```

Comment:
I don't really get the logic here, because you return config if `kind` and `apiVersion` are both None, but if only one is None that will fail.

If I well understand what you want to do is if `kind` or `apiVersion` is `None` just dont check it 
If yes then you need to also check for `None` in the 2 last `if` because one of `apiVersion` or `kind` could be `None`

```suggestion
    if apiVersion and kind:
        return config

    if apiVersion and config['apiVersion'] != apiVersion:
        raise CommandExecutionError(
            'Expected value {} for key apiVersion, got {}'.format(
                config['apiVersion'], apiVersion
            )
        )

    if kind and config['kind'] != kind:
        raise CommandExecutionError(
            'Expected value {} for key kind, got {}'.format(
                config['kind'], kind
            )
        )
```

Also, it's only my opinion but for me it do not hurt to check for empty string or 0 ... for `apiVersion` and `kind` not only for `None`

Reply:
Accepted and will fix after discussing it